packages/contracts/src/contracts/current/forwarder/Forwarder.sol

```diff
@@ -0,0 +1,221 @@
+    }
+
+    /// @dev Buys the tokens, performing fee abstraction if required. This function is payable
+    ///      and will convert all incoming ETH into WETH and perform the trade on behalf of the caller.
+    ///      This function allows for a deduction of a proportion of incoming ETH sent to the feeRecipient.
+    ///      The caller is sent all tokens from the marketBuy of orders (less any fees required to be paid in ZRX)
+    ///      If the purchased token amount does not meet some threshold (98%) then this function reverts.
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param signatures An array of Proof that order has been created by maker.
+    /// @param feeOrders An array of Order struct containing order specifications for fees.
+    /// @param feeSignatures An array of Proof that order has been created by maker for the fee orders.
+    /// @param feeProportion A proportion deducted off the incoming ETH and sent to feeRecipient. The maximum value for this
+    ///        is 1000, aka 10%. Supports up to 2 decimal places. I.e 0.59% is 59.
+    /// @param feeRecipient An address of the fee recipient whom receives feeProportion of ETH.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function buyTokensFee(
+        Order[] orders,
+        bytes[] signatures,
+        Order[] feeOrders,
+        bytes[] feeSignatures,
+        uint16  feeProportion,
+        address feeRecipient)
+        payable
+        public
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        require(msg.value > 0, ERROR_INVALID_INPUT);
+        address token = readAddress(orders[0].takerAssetData, 1);
+        require(token == address(ETHER_TOKEN), ERROR_INVALID_INPUT);
+        require(feeProportion <= MAX_FEE, ERROR_INVALID_INPUT);
+
+        uint256 remainingEthAmount = msg.value;
+        if (feeProportion > 0 && feeRecipient != address(0x0)) {
+            // 1.5% is 150, allowing for 2 decimal precision, i.e 0.05% is 5
+            uint256 feeRecipientFeeAmount = safeDiv(safeMul(msg.value, feeProportion), PERCENTAGE_DENOMINATOR);
+            remainingEthAmount = safeSub(msg.value, feeRecipientFeeAmount);
+            // Transfer the fee to the fee recipient
+            feeRecipient.transfer(feeRecipientFeeAmount);
+        }
+
+        ETHER_TOKEN.deposit.value(remainingEthAmount)();
+        Exchange.FillResults memory fillTokensFillResult = marketBuyTokens(orders, signatures, feeOrders, feeSignatures, remainingEthAmount);
+        addFillResults(totalFillResult, fillTokensFillResult);
+        return totalFillResult;
+    }
+
+
+    function marketBuyTokens(
+        Order[] orders,
+        bytes[] signatures,
+        Order[] feeOrders,
+        bytes[] feeSignatures,
+        uint256 sellTokenAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        uint256 takerTokenBalance = sellTokenAmount;
```

Comment:
Why do we need this to be 2 variables?

Reply:
We modify `takerTokenBalance` when performing fee abstraction to use as the amount to sell in market sell.

`sellTokenAmount` is not always `msg.value` and is kept as is to compare with the final fill results to make sure it's fair.